### ISSUE TITLE
HTTPS access fails with using net/http.

### ISSUE THREAD

**Issue by sho-h** [oneclick/rubyinstaller#353]
HTTPS access fails with using net/http in 2.1.9 and 2.3.3. (Maybe 2.2.6, too)

```
> ruby -r net/http -r uri ^
  -e 'Net::HTTP.get_response(URI.parse("https://downloads.sourceforge.net/project/mingw/MinGW/Extension/bsdtar/basic-bsdtar-2.8.3-1/basic-bsdtar-2.8.3-1-mingw32-bin.zip"))'
...
C:/Ruby21-x64/lib/ruby/2.1.0/net/http.rb:923:in `connect': SSL_connect returned=1 errno=0 state=SSLv3 read server certificate B: certificate verify failed (OpenSSL::SSL::SSLError)
        from C:/Ruby21-x64/lib/ruby/2.1.0/net/http.rb:923:in `block in connect'
        from C:/Ruby21-x64/lib/ruby/2.1.0/timeout.rb:75:in `timeout'
        from C:/Ruby21-x64/lib/ruby/2.1.0/net/http.rb:923:in `connect'
        from C:/Ruby21-x64/lib/ruby/2.1.0/net/http.rb:863:in `do_start'
        from C:/Ruby21-x64/lib/ruby/2.1.0/net/http.rb:852:in `start'
        from C:/Ruby21-x64/lib/ruby/2.1.0/net/http.rb:583:in `start'
        from C:/Ruby21-x64/lib/ruby/2.1.0/net/http.rb:478:in `get_response'
        from -e:1:in `<main>'
```

Maybe, the reason is OpenSSL::X509::DEFAULT_CERT_FILE is invalid.

```
> ruby -r openssl -e 'p OpenSSL::X509::DEFAULT_CERT_FILE'
"C:/Users/Justin/Projects/knap-build/var/knapsack/software/x64-windows/openssl/1.0.1l/ssl/cert.pem"
```

I set SSL_CERT_FILE env var to avoid this in workaround.

```
> ruby -e 'p ENV["SSL_CERT_FILE"]'
"/path/to/cert.pem(from http://curl.haxx.se/ca/cacert.pem)"

> ruby -r net/http -r uri -e 'Net::HTTP.get_response(URI.parse("https://..."))'
(nothing raised)
```

Can we set right value to OpenSSL::X509::DEFAULT_CERT_FILE in distributed binary?

**Comment by sho-h** [comment]
OpenSSL::X509::DEFAULT_CERT_FILE is usable after r4531.

```
--- ext/openssl/ossl_x509.c     (revision 4530)
+++ ext/openssl/ossl_x509.c     (revision 4531)
@@ -13,6 +13,8 @@
 VALUE mX509;

 #define DefX509Const(x) rb_define_const(mX509, #x,INT2FIX(X509_##x))
+#define DefX509Default(x,i) \
+  rb_define_const(mX509, "DEFAULT_" #x, rb_str_new2(X509_get_default_##i()))

 void
 Init_ossl_x509()
@@ -91,5 +93,12 @@
 #if defined(X509_TRUST_OCSP_REQUEST)
     DefX509Const(TRUST_OCSP_REQUEST);
 #endif
+
+    DefX509Default(CERT_AREA, cert_area);
+    DefX509Default(CERT_DIR, cert_dir);
+    DefX509Default(CERT_FILE, cert_file);
+    DefX509Default(CERT_DIR_ENV, cert_dir_env);
+    DefX509Default(CERT_FILE_ENV, cert_file_env);
+    DefX509Default(PRIVATE_DIR, private_dir);
 }
```

X509_get_default_cert_file calls when OpenSSL::X509::DEFAULT_CERT_FILE calls.

X509_get_default_cert_file is below on openssl 1.0.1c.

```
const char *X509_get_default_cert_file(void) {
  return(X509_CERT_FILE);
}
```

And X509_CERT_FILE is below. So OpenSSL::X509::DEFAULT_CERT_FILE is decided on compiling.

```
#define X509_CERT_FILE   OPENSSLDIR "/cert.pem"
```

I have no idea to fix that on RubyInstaller. I close this issue.

**Comment by cabo** [comment]
So did you pursue fixing this at another venue?  Do you have pointers?

This is a real problem.

**Comment by Azolo** [comment]
@cabo This issue actually comes up every so often. It was decided that RubyInstaller wouldn't ship Certificate Authority Certifications,

The workaround is to set the CA Cert File at runtime. The ways I know to do it are per request (which RubyInstaller does), via the Environment variable `SSL_CERT_FILE`, or `OpenSSL::X509::Store.set_default_paths`.

**Comment by cabo** [comment]
Just learned that this has been addressed (since 2.4.1):

https://github.com/oneclick/rubyinstaller2/blob/master/resources/ssl/README-SSL.md

Thanks to all who have kept this issue alive and visible.  And many thanks to @larskanis!
